traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
-        trait = getattr(self, "trait", None)
-        if trait is None:
-            list.extend(self, xlist)
 
-            return
+        Notes
+        -----
+        Notification:
+            index : slice
+                The slice from current_length to new_length if successful.
+            added : list
+                The newly added items.
+            removed : Undefined
+                Will be Undefined
 
-        try:
-            len_xlist = len(xlist)
-        except Exception:
-            raise TypeError("list.extend() argument must be iterable")
+        """
+        index = slice(len(self), len(self) + len(iterable))
+        removed = []
+        added = self.validate(index, removed, iterable)
+
+        super().extend(added)
+
+        self.notify(index, removed, added)
+
+    def insert(self, index, object):
+        """ Insert an object to the list before index.
 
-        if trait.minlen <= (len(self) + len_xlist) <= trait.maxlen:
-            object = self.object()
-            name = self.name
-            validate = trait.item_trait.handler.validate
-            try:
-                if validate is not None:
-                    xlist = [validate(object, name, value) for value in xlist]
+        Parameters
+        ----------
+        index : int
+            The index before which to insert the object.
+        object : Any
+            The object to insert.
 
-                list.extend(self, xlist)
 
-                if (self.name_items is not None) and (len(xlist) != 0):
-                    self._send_trait_items_event(
-                        self.name_items,
-                        TraitListEvent(len(self) - len(xlist), None, xlist),
-                        trait.items_event(),
-                    )
+        Notes
+        -----
+        Notification:
+            index : int
+                The index before which the item was inserted.
+            added : item
+                The newly added items
+            removed : Undefined
+                Will be Undefined
 
-                return
+        """
+        removed = Undefined
+        added = self.validate(index, removed, object)
+        norm_index = self._normalize_index(index)
+
+        super().insert(index, added)
+
+        self.notify(norm_index, removed, added)
+
+    def clear(self):
+        """ Remove all items from the list
 
-            except TraitError as excp:
-                excp.set_prefix("The elements of the")
-                raise excp
 
-        self.len_error(len(self) + len(xlist))
+        Notes
+        -----
+        Notification:
+            index : slice
+                The range of the slice will be 0 to len(self).
+            added : list
+                The empty list [].
+            removed : list
+                The removed items.
+
+        """
+        index = slice(0, len(self), None)
+        removed = [copy.deepcopy(x) for x in self]
+        added = self.validate(index, removed, [])
```

Comment:
Is validation required if we don't intend to use that value?

Reply:
Should we raise an exception or does a warning fit better here? Since we aren't going to use the values returned by the validator when calling `super().clear()`. Are warnings reserved for other kind of behavior in `traits`?

Reply:
My understanding is that the validator has the final say, like a judge, as to how/whether a containment change should take place. If the validator disagrees with the `clear()` method, I am tempted to suggest that `clear()` should not overrule the validator's decision.

Reply:
Resolved in offline discussion, validator will be called in method `clear()`, with removed & added values, however, will not check whether `added == []` for now.